Add transforms to SIMPL

fbudin69500 opened the pull request:
Transforms are added a part of the geomtry object.
Transforms can be simple transforms objects. They contain the following
data:
-Parameters: Parameters describing the transform.
-FixedParameters: Additional parameters, such as the center of rotation.
-TransformTypeAsString: String describing the transform type and its parameters types.
-MovingName: String saving the moving space name.
-ReferenceName: String saving the target space name.

Transforms are saved as part of the DREAM3D HDF5 file.

thewtex commented:
@imikejackson @sdonegan @mgroeber please take a look.

imikejackson commented:
Please run git-clang-format on the new code if possible.

imikejackson commented on Source/SIMPLib/Geometry/CompositeTransform.cpp:
```diff
@@ -0,0 +1,160 @@
+*    United States Air Force Prime Contract FA8650-07-D-5800
+*    United States Air Force Prime Contract FA8650-10-D-5210
+*    United States Prime Contract Navy N00173-07-C-2068
+*
```
Please remove the above 3 contracts and replace with:
 United States Air Force Prime Contract FA8650-15-D-5231

 and in the other files also.

fbudin69500 replied:
done

imikejackson commented on Source/SIMPLib/Geometry/CompositeTransform.cpp:
```diff
@@ -0,0 +1,160 @@
+/* ============================================================================
+* Copyright (c) 2009-2016 BlueQuartz Software, LLC
```
Update the copyright year also. (In all new files)

fbudin69500 replied:
done

thewtex commented on Source/SIMPLib/Common/Constants.h:
```diff
@@ -847,6 +847,18 @@ namespace SIMPL
+    const QString Transform("Transform");
+    const QString CompositeTransform("CompositeTransform");
+    const QString TransformTypeName("TransformTypeName");
+    const QString UnknownTransform("UnkownTransform");
```
`UnknownTransform`

fbudin69500 replied:
done

imikejackson commented on Source/SIMPLib/Geometry/ITransform.h:
```diff
@@ -0,0 +1,56 @@
+    public:
+        SIMPL_SHARED_POINTERS(ITransform)
+        SIMPL_TYPE_MACRO_SUPER_OVERRIDE(ITransform, Observable)
+        virtual int writeTransformToHDF5(hid_t parentId, std::string transformName) = 0;
```
pass strings that are read-only parameters as const std::string &transformName

fbudin69500 replied:
done

imikejackson commented on Source/SIMPLib/Geometry/ITransform.h:
```diff
@@ -0,0 +1,56 @@
+        SIMPL_SHARED_POINTERS(ITransform)
+        SIMPL_TYPE_MACRO_SUPER_OVERRIDE(ITransform, Observable)
+        virtual int writeTransformToHDF5(hid_t parentId, std::string transformName) = 0;
+        virtual int readTransformFromHDF5(hid_t parentId, bool SIMPL_NOT_USED(metaDataOnly), std::string transformName) = 0;
```
pass strings that are read-only parameters as const std::string &transformName

Probably should use the SIMPL_NOT_USED() macro in a CPP file instead of the header? Not sure what the recommended C++11 style is for this.

fbudin69500 replied:
done

imikejackson commented on Source/SIMPLib/Geometry/ITransform.h:
```diff
@@ -0,0 +1,56 @@
+* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#ifndef _itransform_h_
+#define _itransform_h_
```
All the major compilers now support #pragma once statement so we are starting to use that instead of include guards. a C++11 compiler is required to compile SIMPL so this should not be a problem. C++14 is on the way as a requirement.

fbudin69500 replied:
done

Commit:
Add transforms to SIMPL

Transforms are added a part of the geometry object.
Transforms can be simple transforms objects. They contain the following
data:
-Parameters: Parameters describing the transform.
-FixedParameters: Additional parameters, such as the center of rotation.
-TransformTypeAsString: String describing the transform type and its parameters types.
-MovingName: String saving the moving space name.
-ReferenceName: String saving the target space name.

Transforms are saved as part of the DREAM3D HDF5 file.

Filters to convert transforms from and to ITK filters have also been added.

Commit:
ENH: Apply clang-format to Constants.h

Constants.h was modified by PR, but a lot of the changes by clang-format
are touching code that was not modified in this PR, thus this separate
commit.

fbudin69500 commented:
@imikejackson @thewtex Now that I have ran clang-format on Constants.h, the tests are passing :). The PR is ready for review!

imikejackson commented:
I tried rebasing your branch on top of the latest develop and I get a compile error. Does this PR require ITK 5.x?

```
In file included from SIMPLView/SIMPL/SIMPLib/Testing/SIMPLUnitTest.cpp:43:
/Users/mjackson/DREAM3D-Dev/DREAM3D/ExternalProjects/SIMPL/Source/SIMPLib/ITK/Testing/Cxx/itkTransformToDream3DITransformContainerTest.cpp:55:13: error: no matching member function for call to 'SetInput'
    filter->SetInput(itkAffine);
    ~~~~~~~~^~~~~~~~
/Users/mjackson/DREAM3D-Dev/DREAM3D/ExternalProjects/SIMPL/Source/SIMPLib/ITK/itkTransformToDream3DITransformContainer.h:58:16: note: candidate function not viable: no known conversion from 'SmartPointer<itk::AffineTransform<float, 3>>' to 'const SmartPointer<itk::TransformBaseTemplate<float>>' for 1st argument
  virtual void SetInput(const typename ITKTransformType::Pointer& transform);
               ^
/Users/Shared/DREAM3D_SDK/superbuild/ITK/Source/ITK/Modules/Core/Common/include/itkProcessObject.h:524:16: note: candidate function not viable: requires 2 arguments, but 1 was provided
  virtual void SetInput(const DataObjectIdentifierType & key, DataObject *input);
               ^
```

thewtex commented:
Superceded by #231

imikejackson commented:
@fbudin69500 @thewtex Is it possible to get the recent conflicts resolved for this PR?

thewtex commented:
@imikejackson these were merged in #231 so this PR can be closed.

dzenanz commented:
This PR should be closed now.
